### DISCUSSION TITLE
Where can I find the logic for how switch statements are emitted to IL?

### DISCUSSION THREAD

**Discussion by casperOne** [dotnet/roslyn#67937] [Q&A]
I'm currently writing some code to emit IL at runtime for a switch statement (unfortunately, it has to be at runtime since the construction is going to be determined at runtime based on other data only known at runtime).

This switch will be on an integer, which makes it very eligible for a jump table using the switch IL operator.

In looking at decompiled code, I notice the following:

- For branches of two or less items, a conditional is used (check value, branch appropriately, jump)
- For consecutive branches of three or more items which are not too far apart, a jump table is used (filling in with default branch labels to pack the jump table)
- If there are multiple "groups", then a conditional is performed to separate the groups and use conditionals/jump tables as appropriate.

My question is this: where in the Roslyn codebase can I find the logic to determine, given that the input for a switch is an integer, how the logic to determine what IL to emit for a switch statement (whether to branch, use a jump table, etc.).

I've browsed through the code base and haven't had much luck, as well as downloaded the repos and looked through there, but with no luck.

If someone here can point me in the right direction, I'd appreciate it.

Thank you in advance.

**Comment by huoyaoyuan** [comment]
https://github.com/dotnet/roslyn/blob/main/src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_PatternSwitchStatement.cs

This is the entry point of lowering switch statement. I didn't spend time to find the switch decision logic.
